### ISSUE TITLE
[Bug] Android Geofence NullReferenceException

### ISSUE THREAD

**Issue by dahlbyk** [shinyorg/shiny#37]
## Platforms

 - [x] Android

## Feature Library

 - [x] Locations

## Version of Library

1.0.0.287-beta

## Version of OS(s) listed above with issue

Android 6.0.1

## Steps to Reproduce

Not sure honestly, just got a crash report. Presumably "have a geofence"?

## Expected Behavior

No crash.

## Actual Behavior

```
System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object)

System.NullReferenceException: Object reference not set to an instance of an object
  at Shiny.Locations.GeofenceManagerImpl.Process (Android.Content.Intent intent) [0x0003c] in <45153172f6dc477f9ba4a2b27bd23b41>:0 
  at Shiny.Locations.GeofenceBroadcastReceiver+<>c__DisplayClass1_0.<OnReceive>b__0 () [0x0009b] in <45153172f6dc477f9ba4a2b27bd23b41>:0

**Comment by aritchie** [comment]
287 isn’t in public nuget yet.  Please wait for public releases before filing defects on it.  I only work on master so I don’t always have things in order yet

**Comment by aritchie** [comment]
I only work on dev I meant.  So things keep pushing to myget

**Comment by aritchie** [comment]
I couldn't reproduce this.  By crash, are you meaning it is showing up in the logging in the sample?

**Comment by dahlbyk** [comment]
I haven't seen it yet in the Sample (not certain what Shiny build it's using - that would be a neat addition to the Environment screen), but it has now happened twice in the real app I'm close to releasing.

![chrome_2019-07-28_15-21-31](https://user-images.githubusercontent.com/133987/62012434-b2af3080-b14b-11e9-8b85-2a8ebd9ac95c.png)

I do want to correct myself: it's being tracked as an error, not a crash, presumably here: https://github.com/shinyorg/shiny/blob/cfdb474a64d7e85b21bb7206bb7e6fbcaa4f9ffd/src/Shiny.Locations/Platforms/Android/GeofenceManagerImpl.cs#L47-L56

A few comments on that log call in particular:
1. What if each `region.Create()` had its own `try`/`catch` to include region details with the exception?
2. `AggregateException` being a generally unhelpful wrapper around the actual error(s), I wonder about having the Shiny Logging infrastructure check for `AggregateException` and log the `InnerExceptions` directly.
2. Thoughts on having the infrastructure that calls `IShinyStartupTask.Start()` handle the general `try`/`catch`, rather than individual tasks needing to do it?

I need to check with the guy testing the app, but this error may be due to attempting to recreate the geofence after revoking permission for Locations.

**Comment by dahlbyk** [comment]
> 1. What if each `region.Create()` had its own `try`/`catch` to include region details with the exception?

Took a shot at this: https://github.com/shinyorg/shiny/pull/40. I _believe_ this will avoid this particular `AggregateException` as well.

**Comment by aritchie** [comment]
Your NRE is happening during the geofence process.  Is it coming from your delegate code perhaps?  

1. I'm actually going to move this to an individual call instead of an array.
2. based on 1, this won't matter
3. This is by design.  StartupTasks must all pass (unless the author of the task doesn't need them to) as it was designed for wiring up infrastructure.  If they fail, they should crash otherwise you can start seeing missing behavior in your app potentially.

**Comment by dahlbyk** [comment]
> I need to check with the guy testing the app, but this error may be due to attempting to recreate the geofence after revoking permission for Locations.

I can't rule out permissions, but I thought I'd mention I have a _separate_ crash report specifically related to permissions:

<details>
<summary>android.os.Parcel.createException</summary>

> Parcel.java, line 1950
> java.lang.SecurityException: Geofence usage requires ACCESS_FINE_LOCATION permission
>
> android.os.Parcel.createException Parcel.java:1950
> android.os.Parcel.readException Parcel.java:1918
> android.os.Parcel.readException Parcel.java:1868
> com.google.android.gms.internal.zzeb.zzb
> com.google.android.gms.internal.zzbzl.zza
> com.google.android.gms.internal.zzbzu.zza
> com.google.android.gms.internal.zzbzc.zza
> com.google.android.gms.common.api.internal.zzm.zzb
> com.google.android.gms.common.api.internal.zzc.zza
> com.google.android.gms.common.api.internal.zzbr.zzb
> com.google.android.gms.common.api.internal.zzbr.zza
> com.google.android.gms.common.api.internal.zzbp.handleMessage
> android.os.Handler.dispatchMessage Handler.java:102
> android.os.Looper.loop Looper.java:193
> android.os.HandlerThread.run HandlerThread.java:65

</details>

I'd be curious to read your thoughts on the "Shiny way" to handle this, pre-create and if permissions are lost after the geofence is originally added to Shiny's repository.

> Your NRE is happening during the geofence process. Is it coming from your delegate code perhaps?

I'd expect my code to appear in a stack trace if it were, but I doubt it:

```csharp
        private class GeofenceDelegate : IGeofenceDelegate
        {
            public async Task OnStatusChanged(GeofenceState newStatus, GeofenceRegion region)
            {
                Analytics.TrackEvent("Geofence Change", new Dictionary<string, string>
                {
                    { "id", region?.Identifier },
                    { "center", region?.Center?.ToString() },
                    { "radius", region?.Radius?.ToString() },
                    { "status", newStatus.ToString() },
                });

                switch (newStatus)
                {
                    case GeofenceState.Entered:
                        await Api.Post.GeofenceTrigger(region.Identifier, true);
                        break;
                    case GeofenceState.Exited:
                        await Api.Post.GeofenceTrigger(region.Identifier, false);
                        break;
                }
            }
        }
```

**Comment by dahlbyk** [comment]
> Your NRE is happening during the geofence process.

I was looking for where it might have been logged. You're thinking it's in here:

https://github.com/shinyorg/shiny/blob/cfd1ed4ab320fc27008243b79379f10ab256b18f/src/Shiny.Locations/Platforms/Android/GeofenceProcessor.cs#L27

via 

https://github.com/shinyorg/shiny/blob/cfd1ed4ab320fc27008243b79379f10ab256b18f/src/Shiny.Locations/Platforms/Android/GeofenceBroadcastReceiver.cs#L23-L29

https://github.com/shinyorg/shiny/blob/cfd1ed4ab320fc27008243b79379f10ab256b18f/src/Shiny.Core/Platforms/Android/Dispatcher.cs#L11-L21

Odd that the error was reported as `GeofenceManagerImpl.Process(Intent)`.

**Comment by aritchie** [comment]
> Odd that the error was reported as GeofenceManagerImpl.Process(Intent).

The code of where Process was happening has moved in dev which is why you aren't seeing it relating to your error.  

> I'd be curious to read your thoughts on the "Shiny way" to handle this, pre-create and if permissions are lost after the geofence is originally added to Shiny's repository.

The permissions error makes sense and there isn't much I can do about that.  This is the reason I try to blindly restore the geofences on boot.  If permission has been pulled, you'll likely want to do check when your app is in the foreground.  I could downgrade the log to an event instead of an error.  May even add a category specifically around this type of issue  

> I'd expect my code to appear in a stack trace if it were, but I doubt it:

I'm still not seeing where the NRE could be.  Is there more to the stack trace in your appcenter log?

**Comment by dahlbyk** [comment]
> > I'd expect my code to appear in a stack trace if it were, but I doubt it:
> 
> I'm still not seeing where the NRE could be. Is there more to the stack trace in your appcenter log?

There is not. Not sure if that's due to AggregateException or how Dispatcher Task handling works or something else. I can bump to the latest beta to see if this persists.

Beyond #40, which now seems unlikely to be related to what I am seeing, I am not sure how to get a better trace on my actual error. 🤔

I did try revoking Location permission for the sample app with geofences configured…no errors logged so far. 🤷‍♂️

**Comment by dahlbyk** [comment]
Note: I formatted the original exception stack trace as code, as some of the generics (specifically `<OnReceive>`) were being interpreted as HTML. I'm looking closer at the `Process()` code that moved in 1efa4eb1765f367ae48e56a5679d90f13b8c349b, and there are a couple NRE opportunities:

https://github.com/shinyorg/shiny/blob/5cff99d334adb5a3c20633b95eb6dd01409df3cd/src/Shiny.Locations/Platforms/Android/GeofenceManagerImpl.cs#L60-L68

I took a closer look at the [`GeofencingEvent`](https://developers.google.com/android/reference/com/google/android/gms/location/GeofencingEvent) docs and I'm reasonably certain it's an unhandled geofence error. PR incoming (separate from #40).
